### PR TITLE
Support dynamic limit override in ratelimit filter

### LINKED ISSUE #11595: Proposal: support for dynamic rate limits
### Rationale

Wikimedia Foundation is going to adopt envoy as an API Gateway. One of the features we are interested in is rate limiting. In our design, it is a fundamental property of the API that requests are rate limited differently based on who is making the request. Specifically, they stipulate that we will rate limit on a client ID basis, where clients are grouped into classes that share a common limit. Thus, within our system the rate limits are a property of the collection of users. 

Within the current architecture of the ratelimiter service in envoy, it is possible to duplicate the client classes and the limit values into the static config of the rate limiter service, however, duplication of the config creates all kinds of problems, especially if the client classes and their limits are dynamic. 
Instead, we want to have a single source of truth for client properties, including the client classes and rate limits assigned to them, and the natural place to hold this state is user management/authentication system. We can include the limits/classes into the user authentication token via JWT and pass this state into envoy. However, there’s currently no way to pass the state further into the rate limiting service.

We have come up with a proposal solution outlined below, but we’re **very** open to alternatives and seeking guidance on the best way to solve our problem.  We intend to contribute all the code changes.

### Proposal

Expand the [RateLimitDescriptor](https://github.com/envoyproxy/data-plane-api/blob/b742ddcaa09aed7db58adcb29babf4ad2c8f4ecc/envoy/extensions/common/ratelimit/v3/ratelimit.proto#L57) message to optionally include a limit override. Thus, the RateLimiterRequest would be expanded to look like this:

```
# (pseudo-yaml, will be gRPC message, thus it will be backwards-compatible)
RateLimitRequest:
  domain: gateway
  descriptors:
    - value: ( client_id: 1234 )
       limits: # optional
         value: 10
         unit: second
```

Within the [rate limiting service](https://github.com/envoyproxy/ratelimit) the semantics of the override would be the following: if the limit override is provided, it’s treated as if a config matching the descriptor existed in the limiter static configuration. If the override is not provided, the static config is used. Thus, the changes are entirely backwards-compatible and optional.

The [rate limit filter]( https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rate_limit_filter
) in envoy doesn’t need to be changed, essentially it still would still get the list of descriptors and send them into the rate limiter service. 

The [rate limits actions](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-msg-config-route-v3-ratelimit) will have to be changed to support specifying the limit overrides. Strawman config could look somewhat like this:

```
rate_limits:
  - stage: 0
    actions:
      - {request_headers: {header_name: "x-client-id", descriptor_key: "client-id"}}
    limit: { static: { value: 10, unit: seconds } } # optional
```

The limit actions could have various kinds, roughly based on the descriptor action types - we’d be able to provide a static override, take the value/units from headers, or from dynamic metadata.

### PR DESCRIPTION
**Pull request by Pchelolo**
Signed-off-by: Petr Pchelko <ppchelko@wikimedia.org>

Commit Message: Support dynamic limit override in ratelimit filter
Additional Description: Provides a way to specify dynamic rate limit override in the rate limit descriptor from static value or from dynamic metadata. New type, RateLimitUnit was created to share across config protocol and rate limit service protocol. A PR for the reference implementation of the rate limit service will follow after the API changes are discussed and accepted.
Risk Level: low
Testing: unit tests
Docs Changes: Documented new proto messages. Added a section on rate limit overrides to rate_limit_filter.rst
Release Notes:
Fixes #11595

### CODE REVIEW

**Comment by Pchelolo** [comment]
Seems like the changes are hitting the issue in protoc-gen-validate: https://github.com/envoyproxy/protoc-gen-validate/issues/172

File: api/envoy/type/v3/ratelimit_unit.proto [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+option java_package = "io.envoyproxy.envoy.type.v3";
+option java_outer_classname = "RatelimitUnitProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Ratelimit Time Unit]
+
+// Identifies the unit of of time for rate limit.
+enum RateLimitUnit {
```
Comment (mattklein123) [inline api/envoy/type/v3/ratelimit_unit.proto:15]: This is copied out from rls.proto, right? Can we at least put a note in that proto to use this type in the next major version? I think I would also be OK with a compile breaking (not wire) change to just use this new proto for the RLS service but will defer to @htuch on that.

File: docs/root/configuration/http/http_filters/rate_limit_filter.rst [hunk 2]
```diff
@@ -75,6 +75,47 @@ the following descriptor is generated:
...
+
+  actions:
+      - {generic_key: {descriptor_value: some_value}}
+  limit:
+     metadata_key:
+         key: test.filter.key
+         path:
+             - key: test
+
+Will lookup the value of the dynamic metadata, and if it contains the following structure:
```
Comment (mattklein123) [inline docs/root/configuration/http/http_filters/rate_limit_filter.rst:108]: In terms of structure can you be very clear on what the structure needs to be in terms of field names, types, etc.?

Can we also document this here? https://www.envoyproxy.io/docs/envoy/latest/configuration/advanced/well_known_dynamic_metadata and maybe cross link if appropriate?

File: api/envoy/type/v3/ratelimit_unit.proto (hunk 1 above)
Comment (Pchelolo) [reply api/envoy/type/v3/ratelimit_unit.proto:15]: Yeah. I wanted to reuse it there initially and then realized that would not be b/c. Left a TODO comment for v4 in rls.proto

File: docs/root/configuration/http/http_filters/rate_limit_filter.rst (hunk 2 above)
Comment (Pchelolo) [reply docs/root/configuration/http/http_filters/rate_limit_filter.rst:108]: Done.

**Pull request merged**